wasm-libruntime/src/workload.rs

```diff
@@ -0,0 +1,130 @@
+        let engine = wasmtime::Engine::default();
+        let store = wasmtime::Store::new(&engine);
+        let mut module_registry: HashMap<String, wasmtime::Instance> = HashMap::new();
+
+        // Instantiate WASI
+        let wasi = wasmtime_wasi::instantiate_wasi(
+            // preopened_dirs: &[(String, File)],
+            Default::default(),
+            // argv: &[String],
+            Default::default(),
+            // environ: &[(String, String)],
+            Default::default(),
+        )
+        .expect("couldn't instantiate WASI");
+        module_registry.insert(
+            "wasi_unstable".to_string(),
+            wasmtime::Instance::from_handle(&store, wasi),
+        );
+
+        // Resolve the app_module imports using the module_registry.
+        let app_module = wasmtime::Module::new(&store, &app_wasm)?;
+        let imports = app_module
+            .imports()
+            .iter()
+            .map(|import| {
+                let module_name = import.module();
+                if let Some(instance) = module_registry.get(module_name) {
+                    let field_name = import.name();
+                    if let Some(export) = instance.get_export(field_name) {
+                        Ok(export.clone())
+                    } else {
+                        Err(format!(
+                            "Import {} was not found in module {}",
+                            field_name, module_name
+                        ))
+                    }
+                } else {
+                    Err(format!("Import module {} was not found", module_name))
+                }
+                .map_err(Into::into)
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // Invoke the function
+        let app_module_data = wasmtime_interface_types::ModuleData::new(&app_wasm)?;
+        let instance = wasmtime::Instance::new(&app_module, &imports)?;
+        let export = function_name;
+        app_module_data
+            .invoke_export(&instance, export, args)
+            .map_err(Into::into)
+    }
+}
+
+/// Trait to model the functionality for retrieving the WASM workload
+pub trait WorkloadRetriever {
+    fn retrieve(&self) -> Result<Workload>;
+}
+
+/// Retrieve the workload from a raw unux filedescriptor
+pub mod fd_workload_retriever {
+    use super::{Result, Workload};
+
+    use std::{fs::File, io::Read, os::unix::io::FromRawFd};
+
+    /// Default filedescriptor used to read the WASM workload binary
+    const DEFAULT_WORKLOAD_FD: i32 = 3;
+
+    // Struct to implement workload retrieving from a file-descriptor
+    pub struct FdWorkloadRetriever(i32);
+    impl FdWorkloadRetriever {
+        pub fn new(fd: Option<i32>) -> Self {
+            Self(fd.unwrap_or(DEFAULT_WORKLOAD_FD))
+        }
+    }
+
+    impl super::WorkloadRetriever for FdWorkloadRetriever {
+        fn retrieve(&self) -> Result<Workload> {
+            let mut buf = Vec::new();
+            unsafe { File::from_raw_fd(self.0) }
```

Comment:
@haraldh will this work in the kernel you're working on?

Reply:
Ok, something has opened fd 3 for you. Maybe the parent pid, which started you.

But I guess in our case, the bytecode is first transferred over the net. Right?

So, something has to establish the connection and request that payload.

Now, if that something can ensure, that the (maybe ssl transparent) socket has fd 3, then we are good.

Otherwise, we might want to replace `FdWorkloadRetriever` with another `WorkloadRetriever`

@npmccallum ?

Reply:
Scratch my last comment.

Reply:
It will work.. yes.. no problem

Reply:
What I would like to see is a `ReadWorkloadRetriever`
so instead of:
```
let workload_retriever = fd_workload_retriever::FdWorkloadRetriever::new(Some(fd));
```
it will be
```
let workload_retriever = read_workload_retriever::ReadWorkloadRetriever::new(Some(file));
```

with 
pub struct FdWorkloadRetriever<S: Read>(read: S);

because `File` implements Read and &[u8] implements Read... So we can pass a predownloaded slice also.

Reply:
Thanks for the input on the design, which is exactly the remaining item on the TODO list in this PRs description :smiley: I'll tinker with it a bit more!